docs/source/plugins/for_plugin_developers.rst

```diff
@@ -0,0 +1,274 @@
+:func:`~napari.plugins.hook_specifications.napari_get_reader` that enables
+napari to open an imaginary ``.ext`` filetype.
+
+.. code-block:: python
+
+   from pluggy import HookimplMarker
+
+   # we'll get to this line and the decorator below in just a minute
+   napari_hook_implementation = HookimplMarker("napari")
+
+   @napari_hook_implementation
+   def napari_get_reader(path):
+      # remember, path can be a list, so we check it's type first...
+      # (this example plugin doesn't handle lists)
+      if isinstance(path, str) and path.endswith(".ext"):
+         # If we recognize the format, we return the actual reader function
+         return my_reader
+      # otherwise we return None.
+      return None
+
+
+   def my_reader(path):
+      with open(path, 'rb') as file:
+         array = convert_bytes_to_numpy(data)
+      # return it as a list of LayerData
+      return [(array,)]
+
+.. note::
+
+  The seemingly excessive ``list``-of-``tuples`` return type here allows
+  plugins the flexibility of returning multiple layers, with optional
+  layer-construction arguments.
+
+.. _hookimpl-decorator:
+
+Decorating your function with ``pluggy.HookimplMarker``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to let ``napari`` know that one of your functions satisfies the API of
+one of the napari *hook specifications*, you must decorate your function with
+an instance of `pluggy.HookimplMarker
+<https://pluggy.readthedocs.io/en/latest/#marking-hooks>`_, initialized with
+the name ``"napari"``.  (This *does* mean that your plugin needs to depend on
+``pluggy``, but it's a very lightweight dependency that uses only standard lib
+python).
+
+.. code-block:: python
+
+   from pluggy import HookimplMarker
+
+   napari_hook_implementation = HookimplMarker("napari")
+
+
+Currently (as of March, 2020), the only way that napari knows *which* hook
+specification your implementation matches is by looking at the *name* of your
+function.  So in the example above, it was critical that our hook
+implementation was literally named ``napar_get_reader``:
+
+
+.. code-block:: python
+
+   @napari_hook_implementation
+   def napari_get_reader(path: str):
+      ...
+
+However, `a pull request has been merged at pluggy
+<https://github.com/pytest-dev/pluggy/pull/251>`_ that will enable you to mark
+*any* function as satisfying a napari hook specification (regardless of the
+function's name) using the following syntax:
+
+.. code-block:: python
+
+   @napari_hook_implementation(specname="napari_get_reader")
+   def whatever_name_you_want(path: str):
+      ...
+
+(Monitor the `pluggy changelog
+<https://github.com/pytest-dev/pluggy/blob/master/CHANGELOG.rst>`_ for release
+of PR #251.)
+
+.. _plugin-discover:
+
+Step 3: Make your plugin discoverable
+-------------------------------------
+
+Packages and modules installed in the same environment as ``napari`` may make
+themselves "discoverable" to napari using one of two conventions:
```

Comment:
maybe link here to the document you showed us, I think from PyPA (?), when we were first discussing this?